26_Autoexposure/main.cpp

```diff
@@ -0,0 +1,874 @@
+				.meteringWindowOffsetY = MeteringWindowOffset[1] * m_gpuImg->getCreationParameters().extent.height,
+				.lumaMin = LumaMinMax[0],
+				.lumaMax = LumaMinMax[1],
+				.EV = 0.0f,
+				.sampleCountX = SampleCount[0],
+				.sampleCountY = SampleCount[1],
+				.viewportSizeX = m_gpuImg->getCreationParameters().extent.width,
+				.viewportSizeY = m_gpuImg->getCreationParameters().extent.height,
+				.lumaMeterBDA = m_lumaGatherBDA
+			};
+
+			queue->startCapture();
+
+			cmdbuf->begin(IGPUCommandBuffer::USAGE::ONE_TIME_SUBMIT_BIT);
+			cmdbuf->bindComputePipeline(m_lumaMeterPipeline.get());
+			cmdbuf->bindDescriptorSets(nbl::asset::EPBP_COMPUTE, m_lumaMeterPipeline->getLayout(), 0, 1, &ds); // also if you created DS Set with 3th index you need to respect it here - firstSet tells you the index of set and count tells you what range from this index it should update, useful if you had 2 DS with lets say set index 2,3, then you can bind both with single call setting firstSet to 2, count to 2 and last argument would be pointet to your DS pointers
+			cmdbuf->pushConstants(m_lumaMeterPipeline->getLayout(), IShader::E_SHADER_STAGE::ESS_COMPUTE, 0, sizeof(pc), &pc);
+			cmdbuf->dispatch(1 + (SampleCount[0] - 1) / SubgroupSize, 1 + (SampleCount[1] - 1) / SubgroupSize);
+			cmdbuf->end();
+
+			{
+				IQueue::SSubmitInfo submit_infos[1];
+				IQueue::SSubmitInfo::SCommandBufferInfo cmdBufs[] = {
+					{
+						.cmdbuf = cmdbuf
+					}
+				};
+				submit_infos[0].commandBuffers = cmdBufs;
+				IQueue::SSubmitInfo::SSemaphoreInfo signals[] = {
+					{
+						.semaphore = m_lumaMeterSemaphore.get(),
+						.value = m_submitIx + 1,
+						.stageMask = asset::PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT
+					}
+				};
+				submit_infos[0].signalSemaphores = signals;
+
+				queue->submit(submit_infos);
+				queue->endCapture();
+			}
+
+			const ISemaphore::SWaitInfo wait_infos[] = {
+				{
+					.semaphore = m_lumaMeterSemaphore.get(),
+					.value = m_submitIx + 1
+				}
+			};
+			m_device->blockForSemaphores(wait_infos);
+		}
+
+		// Get EV
+		{
+			const auto memory_range = ILogicalDevice::MappedMemoryRange(
+				m_lumaGatherAllocation.memory.get(),
+				0ull,
+				m_lumaGatherAllocation.memory->getAllocationSize()
+			);
+
+			if (!m_lumaGatherAllocation.memory->getMemoryPropertyFlags().hasFlags(IDeviceMemoryAllocation::EMPF_HOST_COHERENT_BIT))
+				m_device->invalidateMappedMemoryRanges(1, &memory_range);
+
+			const uint32_t* buffData = reinterpret_cast<const uint32_t*>(m_lumaGatherAllocation.memory->getMappedPointer());
+
+			assert(m_lumaGatherAllocation.offset == 0); // simpler than writing out all the pointer arithmetic
+
+			m_EV = 0.0f;
+			for (int index = 0; index < SubgroupSize; index++) {
+				m_EV += static_cast<float>(buffData[index]) / (log2(LumaMinMax[1]) - log2(LumaMinMax[0])) + log2(LumaMinMax[0]);
+			}
+			m_EV /= (SampleCount[0] * SampleCount[1]);
```

Comment:
see the discord conversation, the per-workgroup-averages need to be in the `[0,leftoverBits)` range already

then your decode step is
```cpp
for ()
   ev += static_cast<float>(buffData[index]) * (log2(LumaMinMax[1]) - log2(LumaMinMax[0])) / ((1<<fixedPointBitsLeft)-1) + log2(LumaMinMax[0]);
ev /= float(workgroupCount[0]*workgroupCount[1]);
```